jax/_src/clusters/cluster.py

```diff
@@ -0,0 +1,94 @@
+from typing import List, Optional, Union, Sequence, Tuple
+from absl import logging
+from jax._src.cloud_tpu_init import running_in_cloud_tpu_vm
+
+
+class ClusterEnv:
+  """Interface for defining a cluster environment.
+
+  To enable auto bootrapping (aka :func:`jax.distributed.initialize()`),
+  cluster environments need to derive from :class:`ClusterEnv` and implement
+  :func:`is_env_present`, :func:`get_coordinator_address`,
+  :func:`get_process_count`, and :func:`get_process_id`.
+  :class:`ClusterEnv` subclasses are automatically detected when imported.
+  """
+
+  _cluster_types: List['ClusterEnv'] = []
+
+  def __init_subclass__(cls, **kwargs):
+    super().__init_subclass__(**kwargs)
+    cls._cluster_types.append(cls)
+
+  @classmethod
+  def auto_detect_unset_distributed_params(cls,
+                                           coordinator_address: Optional[str],
+                                           num_processes: Optional[int],
+                                           process_id: Optional[int],
+                                           local_device_ids: Optional[Union[int, Sequence[int]]]
+                                          ) -> Tuple[Optional[str], Optional[int], Optional[int],
+                                                     Optional[Union[int, Sequence[int]]]]:
```

Comment:
```suggestion
                                           local_device_ids: Optional[Sequence[int]]
                                          ) -> Tuple[str, int, int, 
                                                     Optional[Sequence[int]]]:
```

Sorry I missed this earlier. These should definitely be set (except local device ids), and we normalize the local devices to a list in distributed.init right?

Reply:
Oh nevermind, they're None if we can't find an environment. I think the local_device_ids should always be a list and never an int though, for both the parameter and returned value.

Reply:
correct, yes it should be none if we don't have an environment. Let me fix the `local_device_ids` type.